core/src/smartcontracts/isi/world.rs

```diff
@@ -122,22 +145,49 @@ pub mod isi {
     impl<W: WorldTrait> Execute<W> for Unregister<Role> {
         type Error = Error;
 
-        #[metrics("unregister_peer")]
+        #[metrics("unregister_role")]
         fn execute(
             self,
             _authority: <Account as Identifiable>::Id,
             wsv: &WorldStateView<W>,
         ) -> Result<(), Self::Error> {
             let role_id = self.object_id;
+
+            let mut accounts_with_role = vec![];
+            for domain in wsv.domains().iter() {
+                let account_ids = domain.accounts().filter_map(|account| {
+                    if account.contains_role(&role_id) {
+                        return Some(account.id().clone());
+                    }
+
+                    None
+                });
+
+                accounts_with_role.extend(account_ids);
+            }
+
+            for account_id in accounts_with_role {
+                wsv.modify_account(&account_id.clone(), |account| {
+                    if !account.remove_role(&role_id) {
+                        error!(%role_id, "role not found - this is a bug");
+                    }
+
+                    Ok(AccountEvent::PermissionRemoved(account_id))
+                })?;
+            }
+
```

Comment:
Good catch! Also I believe it's a good idea to write test when you fix bug. Maybe open a new issue?

Reply:
I've created #2017